File: void_front/game/core/weapons/mine.py
from __future__ import annotations

import numpy as np
from game.core.weapons.shapes import shapes
from game.core.weapons.manager import WeaponManager
from game.core.weapons.minature import Minature
from game.core.weapons.weapon import Weapon


class Mine(Weapon):
    def __init__(self, options, force=False):
        defaults = {
            "name": "Mine",
            "speed": options.get("speed", 10),
            "acceleration": 10000,
            "x": options.get("x"),
            "y": options.get("y"),
            "width": 20,
            "height": 20,
            "angle": options.get("angle"),
            "damage": 50,
            "range": 10000,
            "fireRate": 1,
            "energyCost": getattr(options.get("ship"), "maxHeat", 0) + 100,
            "ship": options.get("ship"),
            "color": options.get("color"),
            "vertices": shapes[0],
            "img": options.get("img"),
        }
        super().__init__(defaults, force)
        self.duration = 1000

    def update(self, dt):
        self.colliding()
        self.duration -= 1
        if self.duration <= 0:
            self.destroy()
            self.explode(200, 1000)

    def explode(self, particles=50, radius=500):
        if not self.active:
            return
        for _ in range(particles):
            prop = {
                "x": self.x - np.sin(self.angle) * self.width,
                "y": self.y - np.cos(self.angle) * self.height,
                "angle": np.random.uniform(-np.pi, np.pi),
                "speed": np.random.uniform(radius / 2, radius),
                "ship": self.ship,
                "range": np.random.uniform(radius * 0.1, radius / 2),
                "color": "yellow",
            }
            WeaponManager.fire(Minature, prop, True)

    def colide(self):
        self.explode()

File: core/weapons/mine.js
import { assets } from "../assets/main.js";
import { randomNum } from "../utils/random.js";
import WeaponManager from "./manager.js";
import Minature from "./minature.js";
import { shapes } from "./shapes.js";
import Weapon from "./weapon.js";

export default class Mine extends Weapon {
    constructor({ x, y, angle, ship, color, speed = 10 }) {
        super({
            name: "Mine",
            speed: speed,
            acceleration: 10000,
            x: x,
            y: y,
            width: 20,
            height: 20,
            angle: angle,
            damage: 50,
            range: 10000,
            fireRate: 1,
            energyCost: ship.maxHeat + 100,
            ship,
            color,
            vertices: shapes[1],
            img: assets?.images?.mine
        });

        this.duration = 1000;
    }

    update() {
        this.colliding();

        if (!(--this.duration)) {
            this.destroy()
            this.explode(200, 1000)
        };
    }


    explode(particles = 50, radius = 500) {
        if (!this.active) return;
        for (let i = 0; i < particles; i++) {
            const prop = {
                x: this.x - Math.sin(this.angle) * this.width,
                y: this.y - Math.cos(this.angle) * this.height,
                angle: randomNum(-Math.PI, Math.PI),
                speed: randomNum(radius / 2, radius),
                ship: this.ship,
                range: randomNum(radius * 0.1, radius / 2),
                color: "yellow"
            }

            WeaponManager.fire(Minature, prop, true)
        }
    }

    colide() {
        this.explode();
    }
}